### ISSUE TITLE
Shared pointer and interfaces

### ISSUE THREAD

**Issue by mathiasi** [emscripten-core/emscripten#9368]
Hi,

First time using emscripten and I've been trying to implement a C++ interface in JavaScript. I have been following the embind docs ( https://emscripten.org/docs/porting/connecting_cpp_and_javascript/embind.html ). The examples work fine but I ran into some issues when I try to combine interfaces and shared pointers.
Consider the following example where I have the _Interface_ that I would like to implement in JavaScript and pass to _foobar_ in C++. If I use a reference it works fine but I would like to use shared pointers.
```C++
#include <emscripten/bind.h>
#include <string>
#include <memory>

using namespace emscripten;

struct Interface {
    virtual ~Interface() = default;
    virtual void invoke(const std::string& str) = 0;
};

struct InterfaceWrapper : public wrapper<Interface> {
    EMSCRIPTEN_WRAPPER(InterfaceWrapper);
    void invoke(const std::string& str) {
        return call<void>("invoke", str);
    }
};

float foobar(int a, int b, std::shared_ptr<Interface> c) {
    c->invoke("From C++");
    return a * b;
}


EMSCRIPTEN_BINDINGS(interface) {
    class_<Interface>("Interface")
        .smart_ptr<std::shared_ptr<Interface>>("Interface")
        .function("invoke", &Interface::invoke, pure_virtual())
        .allow_subclass<InterfaceWrapper>("InterfaceWrapper");

    function("foobar", &foobar);
}
```

```HTML
<!doctype html>
<html>
  <script>
    var Module = {
      onRuntimeInitialized: function() {
        var x = {
            invoke: function(str) {
                console.log('invoking with: ' + str);
            }
        };
        var interfaceObject = Module.Interface.implement(x);
        interfaceObject.invoke("From JavaScript");

        console.log('foobar result: ' + Module.foobar(1,2,interfaceObject));
        
      }
    };
  </script>
  <script src="example.js"></script>
</html>
```
Using
`emcc --bind -o example.js example.cpp`
to compile.
I hope this is even possible and that I could get some guidance. Unfortunately it fails silently right now, with no error-messages in the console.

Thanks!

**Comment by mathiasi** [comment]
After stepping through the code I can tell that the execution ends up with a bindingError: "Passing raw pointer to smart pointer is illegal" (for some reason this doesn't show in the console? I guess it must be some configuration). Is there a way to get `Module.Interface.implement` to return a smart pointer?

**Comment by Niak74** [comment]
Hi, 

I have the same issue, is there a way to create smart pointer of instances from JS-defined classes that derive or implement a C++ interface ?

**Comment by cdrinbublexn** [comment]
Hi,

I've got the same issue, but mine was linked by a call to enable_shared_from_this::shared_from_this() in the parent class, This one require that the instance has an already existing shared_ptr pointing to it somewhere. So I needed the constructor of the derived class on JS side to return a shared_ptr over a raw pointer to avoid a bad weak_ptr exception.

After some digging in Embind's Unit Test, I've found this code sample that make it work in my case.
Don't know if that the solution you seek, though.

https://github.com/emscripten-core/emscripten/blob/2a2c462feee94210b0d9d1052a91dc59f4640cbb/tests/embind/embind_test.cpp#L1431-L1435
